### PR TITLE
GOBBLIN-505: Implement a Git-based FlowGraph Monitor.

### PR DESCRIPTION
**Pull request by sv2000**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-505

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
Create a Git-based FlowGraph monitoring service. This service monitors for changes to a git repo that backs the FlowGraph. The changes include addition/deletion/modification of either a DataNode or FlowEdge file.  

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Unit tests.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java [hunk 1]
```diff
@@ -912,4 +912,18 @@
   public static final String GIT_CONFIG_MONITOR_BRANCH_NAME = GIT_CONFIG_MONITOR_PREFIX + "branchName";
   public static final String DEFAULT_GIT_CONFIG_MONITOR_BRANCH_NAME = "master";
   public static final int DEFAULT_GIT_CONFIG_MONITOR_POLLING_INTERVAL = 60;
+
+  /**
+   * Configuration related to the git flow graph monitoring service
+   */
+  public static final String GIT_FLOWGRAPH_MONITOR_PREFIX = "gitFlowGraphMonitor.";
```
Comment (htran1) [inline gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java]: This set of keys is a copy of the ones above. Can you refactor the logic to share the keys with the defaults encapsulated in the implementations? This allows more of the initialization code to be shared.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+
+  /**
+   * Add an element (i.e., a {@link DataNode}, or a {@link FlowEdge} to
+   * the {@link FlowGraph} for an added, updated or modified node or edge file.
+   * @param change
+   */
+  @Override
+  public void addChange(DiffEntry change) {
+    Path path = new Path(change.getNewPath());
+    if(path.depth() == 3) {
```
Comment (htran1) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Use NODE_FILE_DEPTH?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+      addDataNode(change);
+    } else if(path.depth() == 4) {
```
Comment (htran1) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Use EDGE_FILE_DEPTH?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+
+  /**
+   * Remove an element (i.e. either a {@link DataNode} or a {@link FlowEdge} from the {@link FlowGraph} for
+   * a renamed or deleted {@link DataNode} or {@link FlowEdge} file.
+   * @param change
+   */
+  @Override
+  public void removeChange(DiffEntry change) {
+    Path path = new Path(change.getOldPath());
+    if(path.depth() == 3) {
```
Comment (htran1) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Same comment as above for the values 3 and 4.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    }
+  }
+
+  /**
+   * Add a {@link DataNode} to the {@link FlowGraph}. The method uses the {@link FlowGraphConfigurationKeys#DATA_NODE_CLASS} config
+   * to instantiate a {@link DataNode} from the node config file.
+   * @param change
+   */
+  private void addDataNode(DiffEntry change) {
+    if(checkFilePath(change.getNewPath(), NODE_FILE_DEPTH)) {
```
Comment (htran1) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Please add space after the `if`s in the code. Checkstyle checks for this.

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/core/GitConfigMonitorTest.java [hunk 6]
```diff
@@ -128,7 +128,7 @@ public void cleanUp() {
       this.flowCatalog.stopAsync().awaitTerminated();
     }
 
-    cleanUpDir(TEST_DIR);
+    //cleanUpDir(TEST_DIR);
```
Comment (htran1) [inline gobblin-service/src/test/java/org/apache/gobblin/service/modules/core/GitConfigMonitorTest.java]: Why is this removed?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java (hunk 4 above)
Comment (yukuai518) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Should 3 and 4 be relative to  <root_flowGraph_dir> instead of an absolute path depth?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * Service that monitors for changes to {@link org.apache.gobblin.service.modules.flowgraph.FlowGraph} from a git repository.
+ * The git repository must have an inital commit that has no files since that is used as a base for getting
+ * the change list.
+ * The {@link DataNode}s and {@link FlowEdge}s in FlowGraph need to be organized with the following directory structure on git:
+ * <root_flowGraph_dir>/<nodeName>/<nodeName>.properties
+ * <root_flowGraph_dir>/<nodeName1>/<nodeName2>/<edgeName>.properties
```
Comment (yukuai518) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Do we only support properties extension?  I saw you have hocon files extension as well?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+
+  private FSFlowCatalog flowCatalog;
+  private FlowGraph flowGraph;
+  private String flowGraphDir;
+  private String repositoryDir;
+  private Path flowGraphDirPath;
+  private final PullFileLoader pullFileLoader;
+  private final Config emptyConfig = ConfigFactory.empty();
+  private final Set<String> javaPropsExtensions;
+  private final Set<String> hoconFileExtensions;
```
Comment (yukuai518) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: When both hocon and properties exist in gitRepo, which one will take priority for loading inside of flowCompiler?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java (hunk 7 above)
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Yes, we support ".conf", and ".configuration" as well.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java (hunk 2 above)
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Done.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java (hunk 3 above)
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Done.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java (hunk 4 above)
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Changed to NODE_FILE_DEPTH and EDGE_FILE_DEPTH.

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/core/GitConfigMonitorTest.java (hunk 6 above)
Comment (sv2000) [reply gobblin-service/src/test/java/org/apache/gobblin/service/modules/core/GitConfigMonitorTest.java]: I had removed it for debugging purposes. Forgot to uncomment it back.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java (hunk 5 above)
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Added space after ifs.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java (hunk 8 above)
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Right now, the one that comes later will take the priority i.e. the one that comes later will result in a new FlowEdge that will overwrite the previous FlowEdge.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java (hunk 4 above)
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: @yukuai518 3 and 4 are absolute path depths. We do check relative depths in checkFileRelativeToRoot() too.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java (hunk 7 above)
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Changed the behavior to support only files with ".properties".

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java (hunk 8 above)
Comment (sv2000) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: See comment above. To eliminate non-deterministic behavior introduced by different files with same file names and different file extensions, I have changed the code to support only ".properties" files.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+
+  /**
+   * Helper to check if a file has proper hierarchy.
+   * @param filePath path of the node/edge file
+   * @param depth expected depth of the file
+   * @return true if the file conforms to the expected hierarchy
+   */
+  private boolean checkFileLevelRelativeToRoot(Path filePath, int depth) {
+    Path path = filePath;
+    for (int i = 0; i < depth - 1; i++) {
```
Comment (htran1) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitFlowGraphMonitor.java]: Add null pointer check?
